### DISCUSSION TITLE
AVL result comparison

### DISCUSSION THREAD

**Discussion by cyrildelaporte** [peterdsharpe/AeroSandbox#88] [Q&A] [answered]
Hi, 
I want to do a winglet optimization on Mark Drela's SuperGee glider. 

Before continuing with the actual optimization, I would like to validate the output results from calling AVL in AeroSandbox with results directly from AVL. However, I have not been able to get the same results. I included the .avl file that I have been using here (github doest not accept .avl extension so the .txt should be changed to .avl before using the file) [supergee_no_winglets.txt](https://github.com/peterdsharpe/AeroSandbox/files/10706043/supergee_no_winglets.txt) . 

Running the code below calling AVL out of the box with asb gives me a Cl = 0.255 and Cd = 0.00289. The output directly form AVL gives me Cl = 0.173 and Cd = 0.021, the difference is quite significant. 

For now, I will use the VLM to do the optimization, which actually gives values close to AVL's but it doesn't output the induced drag coefficient, which I want to study.  I have been playing with the ''default_analysis_specific_options" without success. Maybe, someone can give me a tip where I have to look at to make it work.

I am probably overseeing something basic. Apologies for that, I am new to AVL and your code. It looks amazing though, I am looking forward using ASB more in the future. 
```
from aerosandbox import *
# ___________________________________________________________________
# Define Geometry 
sg = Airplane(
    name="SuperGee",
    xyz_ref=[3, 0, 0.5], # CG location
    wings=[
        Wing(
            name="Main Wing",
            xyz_le=[0, 0, 0], # Coordinates of the wing's leading edge
            symmetric=True,
            xsecs=[ # The wing's cross ("X") sections
                WingXSec(  # Root 
                    xyz_le=[0, 0, 0], # Coordinates of the XSec's leading edge, relative to the wing's leading edge.
                    chord=5.5,
                    airfoil=Airfoil(name="naca2412"),
                ), 
                WingXSec(  # Tip
                    xyz_le=[3.5, 14, 0],
                    chord=2,
                    airfoil=Airfoil(name="naca2412")
                ), 
            ]
        ),
    ]
)

# ___________________________________________________________________
# Set Operating Point 
op_point= OperatingPoint(
        Atmosphere(altitude = 6000), 
        velocity= 221.2,
        alpha=0,
        beta=0,
        p=0,
        q=0,
        r=0,)

# ___________________________________________________________________
# Initialize AVL for defined Geometry and Operating Point 
ap= AVL( 
    airplane=sg,
    op_point=op_point
    )

# ___________________________________________________________________
# Run AVL for chose Geometry and Operating Point
output_raw=ap.run() 
for k, v in output_raw.items():
    print(f"{str(k).rjust(10)} : {v}")

```

**Comment by peterdsharpe** [comment]
Hi @cyrildelaporte !

Thanks for the report. Could you please paste your code here using Markdown triple-backtick formatting (e.g., surround your code with ```) so that it can be copy-pasted and reproduced by others?

Guide: https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

**Reply by cyrildelaporte** [reply]
yes that helps, thanks!

**Comment by peterdsharpe** [comment, marked as answer]
Hey @cyrildelaporte !

Awesome - thanks for that! Had a bit of time to dig into the root causes here, and uncovered a few interesting things:

When I run this analysis on my computer, these are the results I see:

|     | AVL 3.37 Result on `supergee_no_winglets.avl` | `asb.AVL` Result on ASB code |
| --- | --------------- | -------------- |
| $C_L$  | 0.173           |      0.255          |
| $C_D$  | *0.00134*         |           0.00288     |

These are mostly consistent with your results, except for the *italicized* result - when I run AVL 3.37 on `supergee_no_winglets.avl` I get a much lower drag coefficient.

I think this is likely a typo in your result, because the results above are consistent with [induced drag theory](https://en.wikipedia.org/wiki/Lift-induced_drag). For example:

$$C_{D,i} = \frac{C_L^2}{\pi (AR) e}$$
Here, $AR=7.47$, and (to first order) the span efficiency $e \approx 1$, so, for the AVL 3.37 result:

$$C_{D,i} = \frac{0.173^2}{\pi \cdot (7.47) \cdot 1} \approx 0.00127$$
Similarly, the `ASB.AVL` result gives a seemingly-correct induced drag coefficient given its (slightly different) lift coefficient.

So, then the question remains - why is `ASB.AVL` giving a slightly different lift coefficient than using AVL directly? It comes down to two things:

### Cause 1: Mach Number Discrepancy

AVL does not take into account the Mach number *if it is specified in the `*.avl` file*, and not set at runtime (via `OPER -> m -> MN {mach number here}`). As far as I can tell, this is a bug in AVL (or, might be the result of deprecated functionality).

In this run file presented, the Mach number is roughly 0.70 (derived from the `OperatingPoint` constructor). The `ASB.AVL` interface takes care of this correctly, by setting this Mach number at runtime. Using `AVL` directly does not - it must be manually set before executing the operating point.

To illustrate this: if I run AVL on `supergee_no_winglets`, I get $C_L=0.173$, same as above. However, if I run AVL on the same file *but first modify the Mach number parameter to M0.7 at runtime*, AVL gives $C_L=0.221$. The resulting lift increase is essentially the result of the Prandtl-Glauert correction happening in AVL.

AeroSandbox's interface to AVL accounts for this nonzero Mach number; using AVL directly does not, unless you specify it at runtime.

### Cause 2: $C_{L\alpha}$ factor
However, this doesn't fully explain the discrepancy - with the corrected Mach number, AVL is still reporting $C_L=0.221$ while `ASB.AVL` reports $C_L=0.255$. 

The remaining difference comes from the presence of a $C_{L\alpha}$ scaling factor in AeroSandbox that AVL lacks (by default), which is explained in the AVL documentation:

```
CLAF        |  (keyword)
CLaf        |  dCL/da scaling factor

This scales the effective dcl/da of the section airfoil as follows:
 dcl/da  =  2 pi CLaf
The implementation is simply a chordwise shift of the control point
relative to the bound vortex on each vortex element.

The intent is to better represent the lift characteristics
of thick airfoils, which typically have greater dcl/da values
than thin airfoils.  A good estimate for CLaf from 2D potential
flow theory is

  CLaf  =  1 + 0.77 t/c

where t/c is the airfoil's thickness/chord ratio.  In practice,
viscous effects will reduce the 0.77 factor to something less.
Wind tunnel airfoil data or viscous airfoil calculations should
be consulted before choosing a suitable CLaf value.

If the CLAF keyword is absent for a section, CLaf defaults to 1.0,
giving the usual thin-airfoil lift slope  dcl/da = 2 pi.
```

Basically, the `supergee_no_winglets.avl` file has a `CLAF` of 1 (yielding a lift-curve slope of $2\pi$). The AVL documentation suggests adding a `CLAF` of around 1.09 given the thickness of the NACA2412 airfoil. 

The AeroSandbox AVL interface does this automatically; using AVL directly does not. Hence, `ASB.AVL` produces a $C_L$ around 9% higher than using AVL directly, which perfectly matches the discrepancy we see.

### Summary

In summary - using `asb.AVL` gives different results than AVL itself, because it's automatically applying "best-practices" corrections that are either a) fixing AVL bugs or b) directly recommended in the AVL documentation. Because of this, **the `ASB.AVL` results here will be the more accurate of the two**.

-----

A final note, regarding this comment:

> For now, I will use the VLM to do the optimization, which actually gives values close to AVL's but it doesn't output the induced drag coefficient, which I want to study

Any vortex-lattice-method aerodynamics code (whether it's AVL, `asb.VortexLatticeMethod`, or otherwise) will only ever give you the induced drag coefficient, as they are fundamentally based on potential flow theory. The drag coefficient reported by `asb.VortexLatticeMethod` is in fact the induced drag (with profile drag of zero), as this is an inherent limitation due to the nature of any VLM code.
